src/backend/distributed/executor/citus_custom_scan.c

```diff
@@ -801,3 +819,92 @@ IsCitusCustomScan(Plan *plan)
 
 	return true;
 }
+
+
+/*
+ * In a Job, given a list of relations, if all them belong to the same
+ * colocation group, the Job's colocation ID is set to the group ID, else,
+ * it will be set to INVALID_COLOCATION_ID.
+ */
+static void
+SetJobColocationId(Job *job)
+{
+	uint32 jobColocationId = INVALID_COLOCATION_ID;
+
+	if (!job->partitionKeyValue)
+	{
+		/* if the Job has no shard key, nothing to do */
+		return;
+	}
+
+	List *rangeTableList = ExtractRangeTableEntryList(job->jobQuery);
+	ListCell *rangeTableCell = NULL;
+	foreach(rangeTableCell, rangeTableList)
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
+		Oid relationId = rangeTableEntry->relid;
+
+		if (!IsCitusTable(relationId))
+		{
+			/* ignore the non distributed table */
+			continue;
+		}
+
+		uint32 colocationId = TableColocationId(relationId);
+
+		if (jobColocationId == INVALID_COLOCATION_ID)
+		{
+			/* Initialize the ID */
+			jobColocationId = colocationId;
+		}
+		else if (jobColocationId != colocationId)
+		{
+			/* Tables' colocationId is not the same */
+			jobColocationId = INVALID_COLOCATION_ID;
+			break;
+		}
+	}
+
+	job->colocationId = jobColocationId;
+}
+
+
+/*
+ * Any function with force_delegate flag(true) must ensure that the Job's
+ * partition key match with the functions' distribution argument.
+ */
+static void
+EnsureForceDelegationDistributionKey(Job *job)
+{
+	/* If the Job has the subquery, punt the shard-key-check to the subquery */
+	if (job->subqueryPushdown)
+	{
+		return;
+	}
+
+	/*
+	 * If the query doesn't have distributed tables, nothing to check, only
+	 * exception is when we have a ShardKey with intermediate_results function
+	 * as the RTE(a subquery plan).
+	 */
+	bool queryContainsDistributedTable =
+		FindNodeMatchingCheckFunction((Node *) job->jobQuery, IsDistributedTableRTE);
+
+	if (!queryContainsDistributedTable && !job->partitionKeyValue)
```

Comment:
seems like we can avoid the expensive FindNodeMatchingCheckFunction check if partitionKeyValue is true

Reply:
Pushed inside the if block